File: src/test/java/com/rafaelfiume/tictactoe/matchers/BoardMatcher.java
package com.rafaelfiume.tictactoe.matchers;

import org.hamcrest.Description;
import org.hamcrest.TypeSafeMatcher;

import static java.lang.String.format;
import static java.lang.System.lineSeparator;

public class BoardMatcher extends TypeSafeMatcher<String> {

    private final String expectedBoard;

    public static BoardMatcher showsABoardLike(String board) {
        return new BoardMatcher(board);
    }

    public BoardMatcher(String expectedBoard) {
        this.expectedBoard = expectedBoard;
    }

    @Override
    protected boolean matchesSafely(String resultBoard) {
        return expectedBoard.equals(resultBoard);
    }

    @Override
    public void describeTo(Description description) {
        description.appendText(format("a board:%s%s", lineSeparator(), expectedBoard));
    }

    @Override
    protected void describeMismatchSafely(String actual, Description mismatchDescription) {
        mismatchDescription.appendText(format("game board was:%s%s", lineSeparator(), actual));
    }
}

File: src/test/java/com/rafaelfiume/tictactoe/matchers/BoardMatcher.kt
package com.rafaelfiume.tictactoe.matchers

import org.hamcrest.Description
import org.hamcrest.TypeSafeMatcher
import java.lang.System.lineSeparator

open class BoardMatcher(private val expectedBoard: String) : TypeSafeMatcher<String>() {

    override fun matchesSafely(resultBoard: String): Boolean = (expectedBoard == resultBoard)

    override fun describeTo(description: Description) {
        description.appendText("a board:${lineSeparator()}$expectedBoard")
    }

    override fun describeMismatchSafely(actual: String, mismatchDescription: Description) {
        mismatchDescription.appendText("board was:${lineSeparator()}$actual")
    }

    companion object {

        @JvmStatic fun showsABoardLike(board: String): BoardMatcher {
            return BoardMatcher(board)
        }
    }

}
